File: app/src/main/java/com/blog/demo/control/viewpager/ViewPagerAdapterActivity.java
package com.blog.demo.control.viewpager;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.blog.demo.LogTool;
import com.blog.demo.R;

import java.util.ArrayList;
import java.util.List;

public class ViewPagerAdapterActivity extends Activity {
    private static final String LOG_TAG = "ViewPagerAdapterActivity";

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_control_view_pager);

        ViewPager viewPager = findViewById(R.id.view_pager);
        viewPager.setAdapter(new DemoPageAdapter(this));

        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset,
                       int positionOffsetPixels) {
                LogTool.logi(LOG_TAG, "position = " + position
                        + ", positionOffset = " + positionOffset
                        + ", positionOffsetPixels = " + positionOffsetPixels);
            }

            @Override
            public void onPageSelected(int position) {
                LogTool.logi(LOG_TAG, "position = " + position);
            }

            @Override
            public void onPageScrollStateChanged(int state) {
                LogTool.logi(LOG_TAG, "state = " + state);
            }
        });
    }

    public static class DemoPageAdapter extends PagerAdapter {
        private List<View> mViewList = new ArrayList<>();

        public DemoPageAdapter(Context context) {
            for (int index = 0; index < 4; index++) {
                TextView view = (TextView) View.inflate(context, R.layout.view_pager_text, null);
                view.setText("第" + index + "页");
                mViewList.add(view);
            }
        }

        // 返回界面数量
        @Override
        public int getCount() {
            return mViewList.size();
        }

        @Override
        public boolean isViewFromObject(View view, Object object) {
            return view == object;
        }

        // 添加界面，一般会添加当前页和左右两边的页面
        @Override
        public Object instantiateItem(ViewGroup container, int position) {
            container.addView(mViewList.get(position));
            return mViewList.get(position);
        }

        // 去除页面
        @Override
        public void destroyItem(ViewGroup container, int position, Object object) {
            container.removeView(mViewList.get(position));
        }

    }

}

File: app/src/main/java/com/blog/demo/control/viewpager/ViewPagerAdapterActivity.kt
package com.blog.demo.control.viewpager

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.blog.demo.LogTool
import com.blog.demo.R

class ViewPagerAdapterActivity : Activity() {
    private val LOG_TAG = "ViewPagerAdapterActivity"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_control_view_pager)

        val viewPager: ViewPager = findViewById(R.id.view_pager)
        viewPager.adapter = DemoPageAdapter(this)

        viewPager.addOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
                LogTool.logi(LOG_TAG, "position = " + position
                            + ", positionOffset = " + positionOffset
                            + ", positionOffsetPixels = " + positionOffsetPixels)
            }

            override fun onPageSelected(position: Int) {
                LogTool.logi(LOG_TAG, "position = $position")
            }

            override fun onPageScrollStateChanged(state: Int) {
                LogTool.logi(LOG_TAG, "state = $state")
            }
        })
    }

    class DemoPageAdapter(context: Context) : PagerAdapter() {
        private val mViewList: MutableList<View> = ArrayList()

        // 返回界面数量
        override fun getCount(): Int {
            return mViewList.size
        }

        override fun isViewFromObject(view: View, obj: Any): Boolean {
            return view === obj
        }

        // 添加界面，一般会添加当前页和左右两边的页面
        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            container.addView(mViewList[position])
            return mViewList[position]
        }

        // 去除页面
        override fun destroyItem(container: ViewGroup, position: Int, obj: Any) {
            container.removeView(mViewList[position])
        }

        init {
            for (index in 0..3) {
                val view = View.inflate(context, R.layout.view_pager_text, null) as TextView
                view.text = "第" + index + "页"
                mViewList.add(view)
            }
        }
    }
}